### PR TITLE
24 keyframe inheritence

### PR DESCRIPTION
**Pull request by mikesimos**
Add ability to stack customizations per environment or environment specific options on top of a core, generic keyframe.
Capitano variadic arguments are used.
connects-to: #24
---- Autogenerated Waffleboard Connection: Connects to #24

### CODE REVIEW

File: src/lib/cli/keyfctl.js [hunk 1]
```diff
@@ -89,18 +89,17 @@ capitano.command({
   }],
   action: (params, options) => {
     if (options.verbose) console.error(options)
-
     const {
       deploy,
       write,
       verbose,
-      keyframe = './keyframe.yml',
+      keyframe = ['./keyframe.yml'],
```
Comment (brownjohnf) [inline src/lib/cli/keyfctl.js:100]: This is setting the default. If someone passes the `-k file.yml` argument when they run it, `keyframe` will be set to a string, not an array. If it's important that it be an array, you'll need to ensure it afterwards.

File: src/lib/models/keyframe.js [hunk 2]
```diff
@@ -6,24 +6,30 @@ const validateSchema = require('jsonschema').validate
...
+      for (let l of kLayers.slice(1)){
+        kf = _.merge(kf, l)
+      }
+      return new Keyframe(kf)
+    })
   }
 
-  constructor(obj, options = {}) {
-    const valid = validateSchema(obj, this.schema())
+  constructor(obj, isCore=false, options = {}) {
```
Comment (brownjohnf) [inline src/lib/models/keyframe.js]: I'd rather put the `isCore` in the options object. That way it avoids modifying the method signature, and since it's technically optional with a default, I think that makes more sense.

File: src/lib/models/keyframe.js [hunk 3]
```diff
@@ -41,12 +47,12 @@ module.exports = class Keyframe {
   }
 
   componentVars(componentName) {
-    return _.get(this.services, `componentName}.variables`, [])
+    return _.get(this.services, [componentName, 'variables'], [])
```
Comment (brownjohnf) [inline src/lib/models/keyframe.js:65]: Oh cool, I didn't realize this supported array syntax.

File: src/lib/models/keyframe.js [hunk 4]
```diff
@@ -41,12 +47,12 @@ module.exports = class Keyframe {
...
   }
 
   services() {
     if (_.has(this, 'services')) return this.services
-    if (_.has(this, 'components')) return this.components
+    if (_.has(this, 'components')) return _.get(this, 'components')
```
Comment (brownjohnf) [inline src/lib/models/keyframe.js:70]: Why this change?

File: src/lib/models/keyframe.js [hunk 5]
```diff
@@ -115,7 +121,20 @@ module.exports = class Keyframe {
...
+  validateKeyframe(obj, isCore=false){
+      if (isCore){
+          return validateSchema(obj, this.coreSchema())
+      }
+      return validateSchema(obj, this.schema())
+  }
+
+  schema(){
+    // Full keyframe schema, including any customization or extension
+    return _.merge(this.coreSchema(), {})
```
Comment (brownjohnf) [inline src/lib/models/keyframe.js:149]: So this will eventually be extended when we know what additional schemas we want to support?

**Review by brownjohnf — CHANGES_REQUESTED** [review]
I think I see what you're doing here, but I think we need to take a slightly different approach. It seems like what we need to do is:

1. Validate the core keyframe against the core schema
2. Merge the add-on files *into* the core keyframe
3. Validate the whole package against the extended schema

That way we can both enforce the presence of certain properties in the core schema, but also ensure that whole, assembled keyframe is compliant with the syntax supported by keyfctl.

I also think we should probably move the schemas to a separate directory and require them at runtime. that way we can extend the schema by dropping additional files into the schemas/ directory.

Does that make sense to you?

File: src/lib/models/keyframe.js [hunk 6]
```diff
@@ -184,7 +201,7 @@ module.exports = class Keyframe {
                     },
                     volumes: {
                       type: 'array',
-                      minItems: 0,
```
Comment (brownjohnf) [inline src/lib/models/keyframe.js:187]: Why are volumes required?

**Comment by brownjohnf** [comment]
Also, I don't see the variadic arguments support in the capitano config.

File: src/lib/models/keyframe.js (hunk 6 above)
Comment (mikesimos) [reply src/lib/models/keyframe.js:187]: they aren't, but It looks clearer to me, that if `volumes:` section is used, it should declare at least one volume. Otherwise, it can be just skipped. 
Should we support this syntax `volumes: {}` ?

File: src/lib/models/keyframe.js (hunk 5 above)
Comment (mikesimos) [reply src/lib/models/keyframe.js:149]: yeah! The `{}` part can have any schema we like and will be merged with the core schema!

File: src/lib/models/keyframe.js (hunk 4 above)
Comment (mikesimos) [reply src/lib/models/keyframe.js:70]: My editor was complaining to be honest. :). Do we have a preference here?

File: src/lib/cli/keyfctl.js (hunk 1 above)
Comment (mikesimos) [reply src/lib/cli/keyfctl.js:100]: I missed that! I tested the default and multiple -k arguments (ie `keyfctl generate -k keyframeCore.yml -k keyframeStaging.yml`). Will fix :+1:

**Comment by mikesimos** [comment]
This was the process more or less, but I wasn't merging and validating to the core keyframe instance but was creating and validating keyframes by merging keyframe.yml layer. The proposed above is a cleaner approach, so followed. Shall we discuss schema input?

File: src/lib/models/keyframe.js (hunk 4 above)
Comment (brownjohnf) [reply src/lib/models/keyframe.js:70]: Woah, very strange. It seems like it's less efficient to use a function from a module (lodash) than a built-in accessor, but it doesn't actually matter in this case. What editor are you using? That's very opinionated. I wonder if @jviotti has any insight/opinion here.

File: src/lib/models/keyframe.js (hunk 6 above)
Comment (brownjohnf) [reply src/lib/models/keyframe.js:187]: Oh, okay. I assumed that if the minItems was set, than you had to have the parent key. If not, then I agree with you.

**Comment by brownjohnf** [comment]
Yeah, let's chat about this when you're back.

File: src/lib/models/keyframe.js (hunk 4 above)
Comment (jviotti) [reply src/lib/models/keyframe.js:70]: The only explanation I can think of from this is that your editor might think that `this` might not be an object, and therefore something ike `this.property` would result in an uncaught exception, and Lodash's `_.get` can handle this edge case, but hard to say. What editor are you using? Can you show us a screenshot of the actual error?

File: src/lib/models/keyframe.js [hunk 7]
```diff
@@ -6,47 +6,68 @@ const validateSchema = require('jsonschema').validate
...
 
-  static fromYaml(yaml) {
-    return new Keyframe(utils.parseYaml(yaml))
+      // validate
+      var valid = kf.validateKeyframeData()
+      if (valid.errors.length > 0) {
+        for (var e of valid.errors){
+          console.error( 'Error: KeyframeLayer.data.components'+ e.stack.slice(8))
+        }
+        process.exit(1)
```
Comment (hedss) [inline src/lib/models/keyframe.js]: Out of interest why a `process.exit()` instead of a `throw`? Usually it's better to throw than exit. Or does something further up trap it? (Same question with the other instances.)

File: src/lib/models/keyframe.js [hunk 8]
```diff
@@ -6,47 +6,68 @@ const validateSchema = require('jsonschema').validate
...
-    if (obj.api_version != 'v2.0.0') {
+    if (obj.api_version !== 'v2.0.0') {
       throw new Error('This version of keyfctl only supports keyframe api version v2.0.0')
     }
 
     this.adapters = _.get(options, 'adapters', {})
 
     if (valid.errors.length > 0) {
-      throw new Error(valid.errors.join('\n'))
+      for (var e of valid.errors){
```
Comment (hedss) [inline src/lib/models/keyframe.js]: Just a comment, but I always shudder when I see single character variable names. ;-)

**Review by hedss — COMMENTED** [review]
Get the idea, looks good, but looks like discussion is needed. :)

File: src/lib/models/keyframe.js (hunk 4 above)
Comment (mikesimos) [reply src/lib/models/keyframe.js:70]: I m using IntelliJ WebStorm 2018.1. 
Components property is available after the class constructor is called with an object with structure 'obj.data.components' at the last line of the constructor ` _.merge(this, obj.data)`. Thus, this warning:
```
Unresolved variable components less... (Ctrl+F1) 
Checks JavaScript referenced variables and fields to be valid ones. The validation works in JavaScript, html or jsp files
```
Shall we keep the change?

File: src/lib/models/keyframe.js (hunk 7 above)
Comment (mikesimos) [reply src/lib/models/keyframe.js]: Throw, yields an (unhandled) exception traceback. I refined error messages a bit, so this makes things look a bit tidier.

File: src/lib/models/keyframe.js (hunk 8 above)
Comment (mikesimos) [reply src/lib/models/keyframe.js]: Shall amend then :)
